lib/spack/spack/solver/asp.py

```diff
@@ -1651,6 +1629,15 @@ def add_concrete_versions_from_specs(self, specs, origin):
                     DeclaredVersion(version=dep.version, idx=0, origin=origin)
                 )
                 self.possible_versions[dep.name].add(dep.version)
+                if (
+                    isinstance(dep.version, spack.version.GitVersion)
+                    and dep.version.user_supplied_reference
+                ):
+                    defined_version = spack.version.Version(dep.version.ref_version_str)
+                    self.declared_versions[dep.name].append(
+                        DeclaredVersion(version=defined_version, idx=1, origin=origin)
+                    )
+                    self.possible_versions[dep.name].add(defined_version)
```

Comment:
Can anybody remind me what is `defined_version`[^1] and explain why it has an hard-coded weight of 1? This is purely for my knowledge and to record the explanation somewhere.

[^1]: I assume it's the `=version` part of a git version, but want to double check.

Reply:
If you have `hash=number` then `defined_version` is the `number`. It has a hard-coded weight of 1 because it is the next possible weight after 0 (i.e. we add either 1 or 2 items in this "dimension" the first having weight 0, and the next one - if added - has weight 1).

Reply:
I don't think this will cause issues right now, but note that this introduces multiple versions with weight 1 in a context where there are multiple versions of weight 0 (and one that is enforced, which is the one above). So, we'll likely have:
```
0 -> this version (which is also enforced by a version_set)
0 -> the default preferred version 
1 -> the defined_version
1 -> the next default preferred version
2 -> ...
```